File: FungeeCommandsManager/src/main/java/org/openredstone/listeners/ChannelListener.java
package org.openredstone.listeners;

import com.google.common.io.ByteArrayDataInput;
import com.google.common.io.ByteStreams;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.event.PluginMessageEvent;
import net.md_5.bungee.api.plugin.Listener;
import net.md_5.bungee.event.EventHandler;
import org.openredstone.FungeeCommandsManager;
import org.openredstone.messages.ReportMessage;

public class ChannelListener implements Listener {
    @EventHandler
    public void onPluginMessage(PluginMessageEvent e) throws Exception {
        if (!e.getTag().equals(FungeeCommandsManager.channel)) {
            return;
        }

        ByteArrayDataInput in = ByteStreams.newDataInput(e.getData());
        String subChannel = in.readUTF();

        if (!subChannel.equalsIgnoreCase(FungeeCommandsManager.subChannel)) {
            return;
        }

        String data = in.readUTF();
        ReportMessage reportMessage = new ReportMessage(data);
        ((ProxiedPlayer) e.getReceiver()).sendMessage(new TextComponent(reportMessage.getColor() + String.join(" ", reportMessage.getArguments()).substring(1)));
    }
}

File: FungeeCommandsManager/src/main/java/org/openredstone/listeners/ChannelListener.kt
package org.openredstone.listeners

import com.google.common.io.ByteStreams
import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.connection.ProxiedPlayer
import net.md_5.bungee.api.event.PluginMessageEvent
import net.md_5.bungee.api.plugin.Listener
import net.md_5.bungee.event.EventHandler
import org.openredstone.messages.ReportMessage

class ChannelListener(val channel: String, val pluginSubChannel: String) : Listener {
    @EventHandler
    fun onPluginMessage(e: PluginMessageEvent) {
        if (e.tag != channel) {
            return
        }
        val byteStream = ByteStreams.newDataInput(e.data)
        val subChannel = byteStream.readUTF()
        if (!pluginSubChannel.equals(subChannel, ignoreCase = true)) {
            return
        }
        val data = byteStream.readUTF()
        val reportMessage = ReportMessage(data)
        (e.receiver as ProxiedPlayer).sendMessage(
            *ComponentBuilder(
                reportMessage.arguments.joinToString(" ").substring(1)
            ).color(reportMessage.color).create()
        )
    }
}
